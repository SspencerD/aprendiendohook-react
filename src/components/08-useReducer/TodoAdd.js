import React from 'react'
import { useForm } from '../../Hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    const [{ description }, handleinputChange, reset] = useForm({
        description: ''

    });


    const handleSubmit =(e)=>{
        e.preventDefault();

        if (description.trim().length <= 1) {

            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddTodo(newTodo);
        reset();

    }

    return (
        <>

            <h4>Agregar</h4>
            <hr />
            <form className="mb-3" onSubmit={handleSubmit}>
                <input
                    className="form-control form-control-sm"
                    type="text"
                    name="description"
                    placeholder="Aprender algo..."
                    autocomplete="off"
                    value={description}
                    onChange={handleinputChange}
                />
                <button
                    className="btn btn-sm btn-outline-success mt-2 btn-block"
                    type="submit"
                >
                    Agregar
                        </button>
            </form>

        </>
    )
}
