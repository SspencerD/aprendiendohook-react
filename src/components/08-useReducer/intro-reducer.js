
const initialState = [{
    id:1,
    todo: 'Estudiar mucho react',
    done: false
}];

const toDo = (state = initialState, action) => {

    if( action?.type=== 'agregar'){
        // como no se ocupa push , se puede ocupar el operador spread
        // ya que llama a la lista anterior y lo agrega
        return [...state , action.payload ];
    }
    //solo puede tener una sola condici√≥n , no se puede agregar un else
    // y otro return ya que estaria en contra del reducer

    return state;
}

let todos = toDo();

// No usar push si estan usando react (reducer)
// ya que modifica  o muta el obj

const newTodo ={
    id:2,
    todo: 'Estudiar mucho redux',
    done: false

}

const addAction ={

    // payload nombre de como se llama al nuevo objeto

    type: 'agregar',
    payload: newTodo
}

todos = toDo(todos, addAction );

console.log(todos);