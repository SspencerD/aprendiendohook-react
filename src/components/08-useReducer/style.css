body{
    padding: 50px;
}

p{
    cursor: pointer;
    margin: 0 px;
}

h1{
    align-items: center;
}

.list-group-item{
    display: flex;
    align-items: center;
    justify-content: space-between;
}
.badge{
    padding-right:  -10px;
    border-radius: 21px;
    cursor: pointer;
}

.complete{
    text-decoration: line-through;
}
