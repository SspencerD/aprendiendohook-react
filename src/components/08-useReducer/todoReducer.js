

export const todoReducer = (state = [], action) => {

    switch (action.type) {
        case 'add':

            return [...state, action.payload];

        case 'delete':

            return state.filter(todo => todo.id !== action.payload); // seria el grupo id

        case 'toggle': //metodo abreviado

        return state.map( todo =>
            ( todo.id === action.payload)
            ? {...todo, done: !todo.done}
            :todo );
        
            case 'toggle-old': //metodo muy largo de hacerlo

            return state.map(todo => {

                if (todo.id === action.payload) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                }else {
                    return todo;
                }

            
            })

        default:
return state;
    }


}