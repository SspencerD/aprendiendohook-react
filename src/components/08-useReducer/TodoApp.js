import React, { useEffect, useReducer } from 'react'
import './style.css';
import { todoReducer } from './todoReducer';
import { Todolist } from './Todolist';
import { TodoAdd } from './TodoAdd';



const init = () => {

    return JSON.parse(localStorage.getItem('todos')) || [];


    // Esto estamos haciendo al momento de renderizar el init
    // return [{
    //     id: new Date().getTime(),
    // desc: 'Aprender React',
    // done: false
    // }];

}
export const TodoApp = () => {


    const [todos, dispatch] = useReducer(todoReducer, [], init);


    useEffect(() => {

        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos]);

    const handleDelete = (todoId) => {

        const borrar = {
            type: 'delete',
            payload: todoId
        };

        dispatch(borrar);

    }

    const handleToggle = (todoId) => {

        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }

    const handleAddTodo = (newTodo) => {

        dispatch({
            type: 'add',
            payload: newTodo

        });
    }



    return (
        <div>

            <h1 className="alert alert-dark">To Do APP <small className="badge bg-info"> {todos.length} </small> </h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <Todolist
                        todos={todos}
                        handleToggle={handleToggle}
                        handleDelete={handleDelete}
                    />
                </div>

                <div className="col-5">

                    <TodoAdd handleAddTodo={ handleAddTodo } />
                   
                </div>
            </div>
        </div>


    )
}
