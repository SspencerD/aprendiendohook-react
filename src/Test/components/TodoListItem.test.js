
import React from 'react';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../components/08-useReducer/TodoListItem';
import { demoTodo } from '../Data/datatodos';


describe('Pruebas en <TodoListItem/>', () => {

    // llamamos las acciones de los botones con jest
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    // incluimos al wapper el componente todolist
    const wrapper = shallow(<TodoListItem
    // incrustamos las propiedades del todolist a la data fake que tenemos
        todo ={ demoTodo[0] }
        index = { 0 }
        handleDelete = { handleDelete }
        handleToggle = { handleToggle }
        />)


    test('debe mostrarse correctamente', () => {
        // hacemos un snap de nuestro codigo

        expect( wrapper).toMatchSnapshot();

        
    });

    test('Debe de llamar la funcion borrar', () => {

        // encontramos la propiedad boton y simulamos el click
        wrapper.find('button').simulate('click');

        // hacemos prueba con la propiedad handledelete y que se encaje
        //con lo que estamos llamando  en este caso a la data fake + id
        expect(handleDelete).toHaveBeenCalledWith(demoTodo[0].id);


        
    });

    test('Debe llamar la función toggle', () => {

        // buscamos el parrafo donde se esta ejerciendo el cambio y simulamos el click
        wrapper.find('p').simulate('click');
        expect(handleToggle).toHaveBeenCalledWith(demoTodo[0].id);
        // estamos llamando  la simulación del click a la data fake + id
        
    });

    test('debe de mostrar el texto correctamente', () => {

        // creamos una variable donde buscamos el parrafo 
        const p = wrapper.find('p');

        //  y esperamos que arroje el valor correctamente al cual estamos llamando
        // en la cual es un texto , borramos los espacio y esperamos que sea igual a data fake
        expect(p.text().trim()).toBe(`1.${demoTodo[0].desc}`);

        //contenido del parrafo
        
    });

    test('debe de tener la clase complete si el todo esta correcto = true', () => {
    

        // hacemos una constante de la data que inicialice en el primer dato
        const todo = demoTodo[0];
        todo.done = true;
        // declaramos que la propiedad done sea true


        //llamamos al wrapper y a la propiedad que requerimos
        const wrapper = shallow(<TodoListItem
                todo ={ todo }
                
                />
                );

                // dentro del obj wrapper llamamos a  propiedad p e indicamos
                // que buscamos tal clase en este caso complete y si este es true.
                expect(wrapper.find('p').hasClass('complete')).toBe(true);
        
    });
    
});