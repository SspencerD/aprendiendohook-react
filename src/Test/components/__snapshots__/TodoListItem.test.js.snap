// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoListItem/> debe mostrarse correctamente 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="'text-center' false"
    onClick={[Function]}
  >
     
    1
    .
    aprender react
  </p>
  <button
    className="btn btn-sm btn-outline-danger mt-1"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
